scrap_emails.py

```diff
@@ -0,0 +1,125 @@
+
+
+    def __init__(self, site='', verify_endpoint=True, *args, **kwargs):
+        """Initializer for the spider that sets up the sites to parse. First uses Request to get the correct base url
+        for a passed site. Then sets any other passed parameters this may choose to implement.
+
+        Args:
+            site: A site to get emails from. In the future, this could be made into a list.
+
+            verify_endpoint: Whether to verify that a redirected link is still under the initial url domain..
+        """
+
+        if not site.startswith('http'):
+            site = 'http://' + site
+
+        r = requests.get(site)
+        if r.status_code == 200:
+            self.start_urls.append(urlparse.urljoin(r.url, '/')[0:-1])
+            self.existing_visited.append(urlparse.urljoin(r.url, '/')[0:-1])
+        else:
+            raise requests.RequestException('Site %s could not be resolved to a HTTP 200 status code' % site)
+
+        # FIXME: Doesn't seem like one can pass a boolean from the command line. Could do like --no-verify as an argument
+        # but easier for now to keep this syntax.
+        if verify_endpoint == "False":
+            self.verify_redirect_endpoint = False
+        else:
+            self.verify_redirect_endpoint = True
+
+        super(EmailSpider, self).__init__(*args, **kwargs)
+
+
+    def parse(self, response):
+        """The main parsing logic of this crawler.
+
+        Args:
+            response: The response from resolving a remote url.
+
+        Returns:
+            A hash of {email: email_value} that is aggregated into an end output.
+        """
+
+        if self.verify_redirect_endpoint and not response.url.startswith(tuple(self.start_urls)):
+            return
+
+        for email in self.get_emails(response.text):
+            yield {'email': email}
+        for link in response.css('a::attr(href)').extract():
+            link = response.urljoin(link)
+
+            #yield {'link': link, 'start_url': self.start_urls[0]}
+
+            # First check as the setting doesn't seem to work as expected in scrapy. Will keep trying to load links after
+            # MAX_DEPTH even if it isn't actually allowing them to process further.
+            if response.meta["depth"] < self.custom_settings['DEPTH_LIMIT'] and self.is_valid_link(link) is True:
+
+                # Add this url to the visited list. Exclude any trailing slash marks for consistency.
+                self.existing_visited.append(re.sub(r'\/$', '', link))
+                yield scrapy.Request(link, callback=self.parse,)
+
+
+    def is_valid_link(self, l):
+        guessed_type = mimetypes.guess_type(l.split('?')[0])
+        # Ensure that we don't parse things like images or pdfs
+        if guessed_type[0] is None or guessed_type[0].startswith('text'):
+            # Ensure that we stick only to this domain.
+            # FIXME: There is an issue with http vs https in this sample application. Also with www vs non-www.
```

Comment:
I should have used sets over arrays since the values are unique. The set lookup is O(1) compared to the O(n) of an array comparison.

There are a few other things that could be done as well. For example to use less memory at the cost of some CPU cycles, one could md5 hash the url for comparison purposes. That would give a consistent 32 characters in memory compared to the [average ~77 characters of
a url](http://www.supermind.org/blog/740/average-length-of-a-url-part-2). 

Furthermore, if one was trying to get all emails from an exceptionally large site where memory usage of the set was a concern, one of two approaches is possible:

1. Write each email or url output to an indexed database field or the an indexer like Solr. This isn't a fast operation but would eliminate the need for a large in-memory set.

2. Split up the subdirectory parsing.
    1. From the index page, create a set of all subdirectories that exist (ie. one for web.mit.edu/faculty, one for web.mit.edu/industry, etc). 
    2. Then process each of those subdirectories individually forcing a match on that base url with that subdirectory. 
    3. If you encounter a new subdirectory that wasn't on the index page, then add it to the list subdirectories to process.
    4. Only process urls that exist in the current subdirectory path.
    5. Once a subdirectory is complete, remove this subdirectory from the "to process" list and clear the set of existing visited urls.
    6. Repeat the process for the next subdirectory.

2b. This second approach essentially splits the site up into smaller site pieces. It does have the disadvantage of potentially missing some emails though. (For example, a link on "faculty"
points to a page on "industry" that isn't linked to from within the "industry" subdirectory. This should be a fairly rare case though).